File: ruby/algorithms/epsilon_greedy/annealing.rb
class AnnealingEpsilonGreedy
  def initialize(n_arms)
    @counts = Array.new(n_arms, 0)
    @values = Array.new(n_arms, 0.0)
  end
  
  def reset(n_arms)
    @counts = Array.new(n_arms, 0)
    @values = Array.new(n_arms, 0.0)
  end

  def select_arm
    t = @counts.reduce(:+) + 1
    epsilon = 1 / Math.log(t + 0.0000001)

    if rand() > epsilon
      @values.index(@values.max)
    else
      rand(@values.size)
    end
  end

  def update(chosen_arm, reward)
     n = @counts[chosen_arm]
     @counts[chosen_arm] = n + 1

     value = @values[chosen_arm]
     if n == 0
       @values[chosen_arm] = reward
     else
       new_value = ((n - 1) / n.to_f) * value + (1 / n.to_f) * reward
       @values[chosen_arm] = new_value
     end
  end
end

File: python/algorithms/epsilon_greedy/annealing.py
import random
import math

def ind_max(x):
  m = max(x)
  return x.index(m)

class AnnealingEpsilonGreedy(object):

  def __init__(self, counts, values):
    self.counts = counts
    self.values = values

  def initialize(self, n_arms):
    self.counts = [0] * n_arms
    self.values = [0.0] * n_arms

  def select_arm(self):
    t = sum(self.counts) + 1
    epsilon = 1 / math.log(t + 0.0000001)

    if random.random() > epsilon:
      return ind_max(self.values)
    else:
      return random.randrange(len(self.values))

  def update(self, chosen_arm, reward):
    new_count = self.counts[chosen_arm] + 1
    self.counts[chosen_arm] = new_count

    value = self.values[chosen_arm]
    new_value = (((new_count - 1) / float(new_count)) * value +
                 (1 / float(new_count)) * reward)
    self.values[chosen_arm] = new_value
